### PR TITLE
Optimized readString for BytesReferenceStreamInput operating on slice array to minimize copying

### PR DESCRIPTION
**Pull request by mosche**
Rather than doing unnecessary copying, iterate over the byte array to identify the number of bytes corresponding to the known number of characters and create a string directly from there. If not enough bytes are available or if usage of surrogate pairs is detected, this aborts and falls back to the default implementation in StreamInput.

Unfortunately writeString ignores surrogate pairs and encodes these naively character by character, which isn't compatible with the UTF-8 encoding of surrogate pairs (a 4-byte sequence).

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java [hunk 1]
```diff
@@ -84,6 +85,60 @@ public long readLong() throws IOException {
...
+            final byte[] bytes = slice.array();
+            final int start = slice.position() + slice.arrayOffset();
+            final int limit = slice.limit() + slice.arrayOffset();
+
+            if (limit - start < chars) {
+                return doReadString(chars); // not enough bytes to read at once
+            }
+
+            // calculate number of bytes matching the requested chars
+            int pos = start;
```
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:102]: This is interesting - I wonder what does it mean wrt `StreamOutput#writeString` (and the [related method I'm adding](https://github.com/elastic/elasticsearch/pull/104680/files#diff-0d24047955e824f5340c51cc91a095c8d3559ed404534c084afb7171e57340dbR462) - which of course needs to be aligned with how we actually write. 

It seems like we are doing it wrong, right? Should we fix it?

Comment (mosche) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:102]: I was wondering about writeString, unfortunately there's no easy fix with regards to backwards compatibility :/
We would have to fix the old readString to properly support 4-byte sequences for surrogates and only 2 (?) major versions later we could start writing these if we're sure they are understood on the other side :/

At this point it doesn't even seem to be worth starting the effort :(

Comment (mosche) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:102]: I think the key is enforcing ASCII whenever possible before even writing.
That allows to get rid of a lot of complicated checks when reading for these cases, such as supporting multiple length characters or even surrogate pairs 🤯

**Review by original-brownbear — APPROVED** [review]
LGTM benchmarked this to be a massive speedup to reading string instances. Lets do it!

**Review by rjernst — COMMENTED** [review]
Can you please add tests? There are a lot of edge cases here that we should ensure work.

File: server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java (hunk 1 above)
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:102]: We can always change the format, it is versioned. But I suggest handling that in a followup.

**Comment by original-brownbear** [comment]
> Can you please add tests? There are a lot of edge cases here that we should ensure work.

@mosche correct me if I'm wrong, but we had all the 2b+/char cases covered by existing tests didn't we? I think the randomisation in the stream input tests covers this already?

File: server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:102]: We should definitely document the deviation from UTF-8, but I suspect we might not want to fix it in the wire protocol. Java itself doesn't really handle the top Unicode plane very well, it's surrogates all the way down, e.g.

```java
assertThat("💩", hasLength(2));
```

If we were to encode top-plane codepoints using the "correct" 4-byte UTF-8 sequences I expect we'd have to introduce quite some extra complexity (e.g. not being able to use `String#length` to count the characters we're sending) and yet I don't see an obvious benefit over what we're doing today. We control both sender and receiver, as long as they agree we can send this stuff however we want.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:102]: > yet I don't see an obvious benefit over what we're doing today

You could build a much faster reader if you had the byte length :) no need for going through the bytes twice. For ASCII encoded strings (which are probably the vast majority) you could just `new String(` on a zero-copy slice out of buffers in most cases (rare to hit Netty buffer boundaries).
The extra cost of writing for us is mostly an implementation detail anyway. If we write a 4 byte length prefix before each string we could do so without iterating the string twice on write or so :)

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:102]: > You could build a much faster reader if you had the byte length :)

Yeah sorry I meant I don't see the benefit of fixing the deviation with UTF-8 w.r.t. surrogate pairs. There's clearly better ways to send strings over the wire. I mean how many of the strings we transmit are things like index names or field names that are already known to both sides, if we leveraged that we could probably avoid instantiating _anything_ most of the time.

**Comment by mosche** [comment]
@elasticmachine update branch

**Comment by mosche** [comment]
>> Can you please add tests? There are a lot of edge cases here that we should ensure work.

> @mosche correct me if I'm wrong, but we had all the 2b+/char cases covered by existing tests didn't we? I think the randomisation in the stream input tests covers this already?

Yes, randomized tests covered this really well. But also happy to add some explicit tests for edge cases if you prefer @rjernst.

File: server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:102]: Is that going to apply to index/field names that are somewhat dynamic? Or is it just for statically-known strings like the name for a `NamedWriteable`?

**Comment by mosche** [comment]
I noticed there's also `ByteBufferStreamInput` and `ByteArrayStreamInput`. These didn't show up in benchmarks, are these used much? If so, it would be fairly simple to adopt the approach to these two as well.

File: server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java (hunk 1 above)
Comment (mosche) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:102]: Just names of NamedWritables for now.

**Comment by mosche** [comment]
Given it was trivial based on the existing work, i've added the zero-copy readString to `ByteBufferStreamInput` and `ByteArrayStreamInput` as well. 

Test coverage is really good, the existing randomized tests reliably surfaced pretty much all corner cases. @rjernst 
```
gradle :server:test --tests "org.elasticsearch.common.*StreamInputTests" -Dtests.iters=100
gradle :server:test --tests "org.elasticsearch.common.io.stream.*" -Dtests.iters=100
```

File: server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java [hunk 2]
```diff
@@ -84,6 +85,23 @@ public long readLong() throws IOException {
...
+        final int chars = readArraySize();
+        if (slice.hasArray()) {
+            // attempt reading bytes directly into a string to minimize copying
+            final byte[] bytes = slice.array();
+            final int start = slice.position() + slice.arrayOffset();
+            final int limit = slice.limit() + slice.arrayOffset();
+            final int length = calculateByteLengthOfChars(bytes, chars, start, limit);
+            if (length >= 0) {
+                slice.position(length + start - slice.arrayOffset());
+                return new String(bytes, start, length, StandardCharsets.UTF_8);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:99]: If `length == chars` (surely the most common case) I think it's safe to use `ISO_8859_1` instead of `UTF_8`, which skips a few extra checks when constructing the `String`.

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:99]: In fact this is such a common case that I wonder if it'd be even faster just to scan `bytes` looking to see if any of them have the top bit set, falling back to today's more general impl otherwise. Seems like an easier thing for the JIT to optimise into a really tight loop vs all the branching in `calculateByteLengthOfChars`.

Comment (mosche) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:99]: @original-brownbear WDYT, we've talked about this. Happy to come up with another impl so we can benchmark and compare. In any case, it'll be a lot simpler!

Comment (ldematte) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java:99]: Nice idea! I agree with @mosche, if we want to try it this is worth a benchmark :)
I wonder if we could just use jmh here, since what we want to benchmark is pretty well defined (do we use it anywhere in the codebase?)

Comment (mosche) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java]: > In fact this is such a common case that I wonder if it'd be even faster just to scan bytes looking to see if any of them have the top bit set, falling back to today's more general impl otherwise. Seems like an easier thing for the JIT to optimise into a really tight loop vs all the branching in calculateByteLengthOfChars.

@DaveCTurner I've optimized the loop as suggest, let me know what you think.

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 3]
```diff
@@ -523,30 +526,44 @@ protected String doReadString(final int charCount) throws IOException {
...
+        if (limit < end) {
+            return null; // not enough bytes to read chars
         }
-        // calculate number of bytes matching the requested chars
-        int pos = start;
-        int remaining = chars;
-        while (pos < limit && remaining-- > 0) {
+        for (int pos = start; pos < end; pos++) {
+            if ((bytes[pos] & 0xf0) > 0x70) {
+                // not an ISO_8859_1 char, fall back to reading a UTF-8 string
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:536]: Nit: `ISO_8859_1` includes characters with the top bit set, we're just checking for `US-ASCII` here.

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 4]
```diff
@@ -523,30 +526,44 @@ protected String doReadString(final int charCount) throws IOException {
...
+        final int end = start + chars;
+        if (limit < end) {
+            return null; // not enough bytes to read chars
         }
-        // calculate number of bytes matching the requested chars
-        int pos = start;
-        int remaining = chars;
-        while (pos < limit && remaining-- > 0) {
+        for (int pos = start; pos < end; pos++) {
+            if ((bytes[pos] & 0xf0) > 0x70) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:535]: I'm curious, is this better than `(bytes[pos] & 0x80) != 0` in some way? Deserves a comment I think.

**Review by DaveCTurner — COMMENTED** [review]
Looks good, a couple of nonblocking comments.

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 4 above)
Comment (mosche) [reply server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:535]: that just originated from `(bytes[pos] & 0xff) >> 4 > 7` without much thinking, just checking the msb is obviously simpler. I'll change, thanks :)

**Comment by mosche** [comment]
@elasticmachine update branch

**Comment by mosche** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 5]
```diff
@@ -531,6 +537,58 @@ public String readString() throws IOException {
...
+            return null; // not enough bytes to read chars
+        }
+        for (int pos = start; pos < end; pos++) {
+            if ((bytes[pos] & 0x80) != 0) {
+                // not an ASCII char, fall back to reading a UTF-8 string
+                return tryReadUtf8StringFromBytes(bytes, start, limit, pos, end - pos);
+            }
+        }
+        skip(chars); // skip the number of chars (equals bytes) on the stream input
+        return new String(bytes, start, chars, ISO_8859_1);
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:552]: shouldn't this encoding be ASCII? ISO-8859-1 is not a subset of the encoding for utf8 (the top bit may be set). Effectively it won't matter since ASCII is a subset of ISO-8859-1, but we should set the right encoding.

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 6]
```diff
@@ -531,6 +537,58 @@ public String readString() throws IOException {
...
+
+    private String tryReadUtf8StringFromBytes(final byte[] bytes, final int start, final int limit, int pos, int chars) throws IOException {
+        while (pos < limit && chars-- > 0) {
+            int c = bytes[pos] & 0xff;
+            switch (c >> 4) {
+                case 0, 1, 2, 3, 4, 5, 6, 7 -> pos++;
+                case 12, 13 -> pos += 2;
+                case 14 -> {
+                    // surrogate pairs are incorrectly encoded, these can't be directly read from bytes
+                    if (maybeHighSurrogate(bytes, pos, limit)) return null;
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:563]: should this be an exception, like the fallback throwOnBrokenChar?

**Review by rjernst — APPROVED** [review]
LGTM. I have a couple last asks.

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 5 above)
Comment (mosche) [reply server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:552]: @DaveCTurner brought this up, using `ISO-8859-1` safes some additional checks compared to `UTF_8`, but also a few compared to `US_ASCII`. By this line we've already check each byte is within [0, 127], so `ISO-8859-1` is safe to use and internally the coder is `LATIN1` in either case.

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 6 above)
Comment (mosche) [reply server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:563]: `null` triggers the fallback `doReadString` (using additional buffers), that one can handle surrogate pairs in our own custom encoding.

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 5 above)
Comment (mosche) [reply server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:552]: Oh, I missed your comment @DaveCTurner . I'll add a comment on that line 👍
